sys/net/gnrc/sock/ip/gnrc_sock_ip.c

```diff
@@ -195,7 +195,26 @@ ssize_t sock_ip_send(sock_ip_t *sock, const void *data, size_t len,
     if (res <= 0) {
         return res;
     }
+#ifdef SOCK_HAS_ASYNC
+    if ((sock != NULL) && (sock->reg.async_cb.ip)) {
+        sock->reg.async_cb.ip(sock, SOCK_ASYNC_MSG_SENT);
+    }
+#endif  /* SOCK_HAS_ASYNC */
     return res;
 }
 
+#ifdef SOCK_HAS_ASYNC
+void sock_ip_set_cb(sock_ip_t *sock, sock_ip_cb_t cb)
+{
+    sock->reg.async_cb.ip = cb;
+}
+
+#ifdef SOCK_HAS_ASYNC_CTX
+sock_async_ctx_t *sock_ip_get_async_ctx(sock_ip_t *sock)
+{
+    return &sock->reg.async_ctx;
+}
```

Comment:
The function is part of the `sock_async` API so its name is already merged. However

> Code looks good, just two small things.
> The name of `sock_ip_get_async_ctx()` is confusing. It's not a getter for `async_ctx` but a general access wrapper.
> Not sure if it is even necessary, just makes the code slightly nicer to read. But then it might as well be `static inline` so the nicer syntax doesn't come at a cost.

As far as I know this *is* a getter. So I don't know why the name is confusing.

> Since this is just a 'nicer' way to access a field of the `sock_ip_t`, it should better be a `static inline` function so the compiler can optimize it away.

`sock_ip_t` is defined by the network stack, so different for any network stack. An inline is thus not easily possible.

Reply:
So is the idea to set `async_ctx` by calling `*sock_ip_get_async_ctx(sock) = ctx;` ?
And where is `async_ctx` defined to begin with?

Reply:
(Sidenote: maybe its good its confusing, so the user is not tempted to use something he is not intended to, but is supposed to be used by the frontend like #12601 or #12602 ;-P

Reply:
> So is the idea to set `async_ctx` by calling `*sock_ip_get_async_ctx(sock) = ctx;` ?
> And where is `async_ctx` defined to begin with?

You don't set the context. It is just a helper for the frontend to have a storage type within the sock and get it from the sock stack-indepent. Have a look at #12601 and #12602. For #12601 `async_ctx` is just a `kernel_pid_t` and `sock_ip_get_async_ctx()` returns a pointer to that position within the `sock` object (so setting happens by dereferencing the pointer). For #12602 it is the event + pointer to the queue, so setting just happens by providing values for the members.

Reply:
I see, so this is not to be used/useful on it's own, but part of a larger feature.

I'd drop the `#ifdef SOCK_HAS_ASYNC_CTX` bits as those are never used and wouldn't even compile right now, but I see how it's convenient to not have that commit in two PRs.

Reply:
> I'd drop the `#ifdef SOCK_HAS_ASYNC_CTX` bits as those are never used and wouldn't even compile right now, but I see how it's convenient to not have that commit in two PRs.

I don't get what you mean by that. What do you propose to do if the front-end does not need a context?

Reply:
It's not about the `#ifdef`, it's that `sock_async_ctx_t` is not defined anywhere in current master or this PR.

Reply:
Ah, so you miss the implementation ;-P